[LowerMemIntrinsics] Optimize memset lowering

ritter-x2a opened the pull request:
This patch changes the memset lowering to match the optimized memcpy lowering.
The memset lowering now queries TTI.getMemcpyLoopLoweringType for a preferred
memory access type. If that type is larger than a byte, the memset is lowered
into two loops: a main loop that stores a sufficiently wide vector splat of the
SetValue with the preferred memory access type and a residual loop that covers
the remaining bytes individually. If the memset size is statically known, the
residual loop is replaced by a sequence of stores.

This improves memset performance on gfx1030 (AMDGPU) in microbenchmarks by
around 7-20x.

I'm planning similar treatment for memset.pattern as a follow-up PR.

For SWDEV-543208.

ritter-x2a committed:
[LowerMemIntrinsics] Optimize memset lowering

This patch changes the memset lowering to match the optimized memcpy lowering.
The memset lowering now queries TTI.getMemcpyLoopLoweringType for a preferred
memory access type. If that type is larger than a byte, the memset is lowered
into two loops: a main loop that stores a sufficiently wide vector splat of the
SetValue with the preferred memory access type and a residual loop that covers
the remaining bytes individually. If the memset size is statically known, the
residual loop is replaced by a sequence of stores.

This improves memset performance on gfx1030 (AMDGPU) in microbenchmarks by
around 7-20x.

I'm planning similar treatment for memset.pattern as a follow-up PR.

For SWDEV-543208.

ritter-x2a commented:
One thing that's not ideal with this patch is that the lowering for memsets with large statically known sizes wastes registers with the SelectionDAG ISel:
In this case, the IR lowering uses a `<64xi32>` store in the main memset loop.  We correctly legalize this into 16 dwordx4 stores, but the huge `<64xi32>` splat value that is stored there (and which consists of 256 times the same byte) lives in a different basic block than the stores. SDAG ISel therefore doesn't know that those 64 32-bit registers with the same value are not needed at the same time and that 4 would be enough (GlobalISel, since it can look across BBs, doesn't have this problem).

You can see this for example in `@memset_p0_sz1055_align_4_varsetval` in `memset-param-combinations.ll`.

I tried adjusting the IR lowering to put the splat values in the same basic block as the accesses, but then they are LICM-ed out again.

I also tried adjusting the lowering to use N(=16) `<4xi32>` stores (with only a single `<4xi32>` splat), and while that fixed the register wastage, it made code generation worse in a different way, because the SCEV-based strength reduction (loop-reduce) then replaces the address computations with new computations that don't use `inbounds` and `nuw`, which means that offsets cannot be folded into store instructions in various cases. This even happens if I change the memset lowering to produce the form that loop-reduce would generate: it still re-generates the address computation minus the poison-generating flags.

The effect of the register wastage here is probably in practice not very dramatic because this only happens for quite large memsets that will take some time anyway, but do let me know if you have suggestions on how to avoid it.

krzysz00 reviewed (commented):
Can't see any issue on the buffer fat pointer side, though it's weird that all those casts and extracts didn't fold

ritter-x2a committed:
Add an overload of `expandMemSetAsLoop` that takes an optional TTI pointer

This avoids breaking the API for out-of-tree tools like the
SPIRV-LLVM-Translator.

arsenm commented on llvm/lib/Transforms/Utils/LowerMemIntrinsics.cpp:
```diff
@@ -983,10 +983,197 @@ static void createMemMoveLoopKnownSize(Instruction *InsertBefore,
+                                Value *SetValue, Type *DstType) {
+  unsigned DstSize = DL.getTypeStoreSize(DstType);
+  Type *SetValueType = SetValue->getType();
+  unsigned SetValueSize = DL.getTypeStoreSize(SetValueType);
```
Keep these all in TypeSize

ritter-x2a committed:
Reviewer feedback: use TypeSize for type sizes, drop unnecessary bitcast
type check, properly format call argument comments.

ritter-x2a commented on llvm/lib/Transforms/Utils/LowerMemIntrinsics.cpp:
```diff
@@ -983,10 +983,197 @@ static void createMemMoveLoopKnownSize(Instruction *InsertBefore,
+  // The value has the right size, but we might have to bitcast it to the right
+  // type.
+  if (Result->getType() != DstType) {
+    Result = B.CreateBitCast(Result, DstType, "setvalue.splat.cast");
```
Done.

ritter-x2a commented:
> I almost want to make TTI mandatory, but these are standalone utility functions. So it may be slightly useful to support them without TTI 

Agreed on both points. So, I'd suggest keeping this part as it is in this PR. Should I change the memcpy and memmove lowerings to be consistent in a follow-up PR, or do you think it makes more sense to add those overloads when/if they are needed?

ritter-x2a committed:
Update the new Transforms/PreISelIntrinsicLowering/AMDGPU/expand-mem-intrinsics.ll test to the new memset lowering.
